machida/lib/wallaroo/experimental/__init__.py

```diff
@@ -100,6 +125,346 @@ def write(self, message, event_time=0, key=None):
+            new = Stream(stream_id,
+                         stream_name,
+                         0 if point_of_ref is None else point_of_ref,
+                         False)
+        # update locally
+        self._streams[stream_id] = new
+        # send to wallaroo worker
+        self.write(cwm.Notify(new.id,
+                              new.name,
+                              new.point_of_ref))
+
+    def end_of_stream(self, stream_id, event_time=None, key=None):
+        """
+        Send an EOS message for a stream_id, with an optional key and
+        event_time.
+        event_time must be either a datetime or a float of seconds since
+        epoch (it may be negtive for dates before 1970-1-1)
+        """
+        flags = cwm.Message.Eos | cwm.Message.Ephemeral
+        if event_time is not None:
+            flags |= cwm.Message.EventTime
+            if isinstance(event_time, datetime):
+                ts = dt_to_timestamp(event_time)
+            elif isinstance(event_time, (int, float)):
+                ts = event_time
+            else:
+                raise ProtocolError("Event_time must be a datetime or a float")
+        else:
+            ts = None
+        if key:
+            flags |= cw.Message.key
+            if isinstance(key, bytes):
+                en_key = key
+            else:
+                en_key = key.encode()
+        else:
+            en_key = None
+        msg = cwm.Message(
+            stream_id = stream_id,
+            flags = flags,
+            message_id = None,
+            event_time = ts,
+            key = en_key,
+            message = None)
+        self.write(msg)
+
+    ########################
+    # User defined methods #
+    ########################
+
+    def stream_updated(self, stream):
+        """
+        User handler can handle updates to their streams here.
+        Updates may include:
+            - stream has been closed (Stream.is_open = False)
+            - stream has been open (Stream.is_open = True)
+            - point of reference has been updated (either forward or back!)
+        """
+        pass
+
+    def handle_invalid_message(self, msg):
+        print("WARNING: {}".format(ProtocolError(
+            "Received an unrecognized message: {}".format(msg))))
+        pass
+
+    def handle_restart(self):
+        # TODO:
+        # 1. close the connection
+        # 2. start new handshake
+        # 3. notify all current streams
+        #    cannot assume that any messages sent with message_id's larger
+        #    than those last acked (e.g. last point_of_ref in self._streams)
+        #    have been processed.
+        # TODO: should we reset streams to last acked point, or wait for the
+        #       worker to respond to the notify acks? Or at least to the `Ok`?
+        #       What if once reconnected, not all streams in self._streams`
+        #       show up in the Ok message?
+        print("Got RESTART... now what?")
```

Comment:
This should work the same as an error except that the server requested the restart so perhaps if we do any retry limits or backoff, we can treat this separately since it's likely a cluster state recovery rather than anything specifically wrong with the operation of the session.